core/src/mast/serialization/basic_block_data_decoder.rs

```diff
@@ -55,7 +57,12 @@ impl<'a> BasicBlockDataDecoder<'a> {
             } else {
                 // decorator.
                 let decorator = self.decode_decorator(&mut data_reader)?;
-                decorators.push((operations.len(), decorator));
+                let decorator_id = mast_forest.add_decorator(decorator).map_err(|err| {
+                    DeserializationError::InvalidValue(format!(
+                        "failed to add decorator to MAST forest: {err}"
+                    ))
+                })?;
+                decorators.push((operations.len(), decorator_id));
```

Comment:
Ah yes, good catch. But we just need to use `MastForestBuilder` in deserialization instead of `MastForest`, which would fix the problem.

This also made me notice that we are not serializing the decorators of non-basic block nodes... Will fix that too.

Reply:
Actually I thought more about it and using `MastForestBuilder` in deserialization might not be a good idea, since using a different version of the builder when deserializing can lead to differences in `MastNodeIds`, and it's probably best to avoid that. We could ensure with a version string that the builder at deserialization time is compatible with the one used at serialization time, but that seems overly complex and bug-prone.

So I think it's worth having larger debug serialized forests using your suggestion (in the next comment) than more compact ones (from the current implementation).

Reply:
Here is my planned changes that should address this comment (and the other related ones):

- Decorators serialized similar to `MastNode`s
    - i.e. store a `decorators: Vec<DecoratorInfo>`
    - in the forest, `DecoratorId`s point into this buffer
- Change how `BasicBlockNode` is serialized
    - decorators are no longer stored interleaved with operations
    - `MastNodeType::Block` redefined as:
```rs
enum MastNodeType {
    // ...
    Block {
        // Vec<Operation> offset
        offset_ops_offset: u32,
        // offset of `DecoratorList` in `data` buffer, or 0
        offset_decorators_offset: u32
    }
    // ...
}
```

- All non-basic block nodes will not store their before/after decorators directly in `MastNodeInfo`
    - instead, we store these as 2 maps at the forest level (serialized):
    1. `before_enter: BTreeMap<MastNodeId, Vec<DecoratorId>>`
    2. `after_exit: BTreeMap<MastNodeId, Vec<DecoratorId>>`

Reply:
Makes sense! A couple of questions/comments:

1. What is `DecoratorInfo`? Is it the same as `Decorator`?
2. Why do we say `offset_ops_offset` and `offset_decorators_offset` rather than just `offset_ops_offset` and `decorators_offset`?
3. For the basic block, do we not need to also specify the number of operations and the number of decorators? (i.e., are just offsets enough)?

Reply:
Addressed in #1482